### DISCUSSION TITLE
How to Properly Utilize Values Class

### DISCUSSION THREAD

**Discussion by mrname** [sqlalchemy/sqlalchemy#8292] [Usage Questions]
I am working with postgres, and trying to create a query along the lines of `UPDATE ... FROM VALUES(... values list ...)`. Essentially, exactly what is shown in [this wiki article](https://github.com/sqlalchemy/sqlalchemy/wiki/PGValues#pgvalues).

I can see that since the wiki article was originally created, support for `Values` has been [directly added into core](https://github.com/sqlalchemy/sqlalchemy/blob/main/test/sql/test_values.py). However, I was unable to find any reference to this in the documentation.

I would be happy to prepare a pull request adding the docs but I was unable to get it working properly for update statements. The basic examples from these tests do indeed work for selects, but not updates, at least not in the various incantations I attempted. Any guidance would be appreciated

Thanks in advance for the assistance!

**Comment by CaselIT** [comment]
Hi,

The documentation is here https://docs.sqlalchemy.org/en/14/core/selectable.html?highlight=values#sqlalchemy.sql.expression.Values and also in the `values()` function 

Currently it's only available as a from event, so you cannot use it in other places, line in an IN clause or similar.
In the update example it seems to be used as a from so it should be working. Could you provide a runnable example so that we can debug the error?

**Reply by mrname** [reply]
Thanks for the quick response! When you say "In the update example" I assume you mean the one from the wiki (the `Values` docs contain only an example for `select`). It is unclear to me exactly how I should be applying the `FROM` clause to the update statement, so here is an example of what I am trying right now:

```
import sqlalchemy as sa

from sqlalchemy import update, values
from sqlalchemy.orm import declarative_base

from .session_factory import db_session

Model = declarative_base()

class User(Model):
    id = sa.Column(Integer, primary_key=True)
    name = sa.Column(sa.String())

user = User(name='test')                                           
db_session.add(user)                                                                                                         
db_session.commit()

value_expr_for_update = values(                                                                                             
    sa.column('id', sa.Integer),                                                                                                  
    sa.column('name', sa.String),                                                                                                 
    name="my_values"                                                                                                        
).data(                                                                                                                     
    [(user.id, 'new_name')]                                                                                                  
)

# Using ORM, as shown in Wiki
db_session.query(User).filter(User.name == 'test').update(dict(name=value_expr_for_update.c.name))

# Or, using core update statement
update_statement = update(User.__table__).values(name=value_expr_for_update.c.name).where(User.name == 'test') 
db_session.execute(update_statement)
```

Both invocations result in the same exception:

```
sqlalchemy.exc.ProgrammingError: (psycopg2.errors.UndefinedTable) missing FROM-clause entry for table "my_values"
LINE 1: UPDATE user SET name=my_values.name WHERE user.name = 'test'
```

Additionally, the ORM usage emits a warning:

```
SAWarning: Evaluating non-mapped column expression 'my_values.name' onto ORM instances; this is a deprecated use case.   
Please make use of the actual mapped columns in ORM-evaluated UPDATE / DELETE expressions.
```
